### ISSUE TITLE
BackgroundJobClient.ChangeState method fails being nested in TransactionScope

### ISSUE THREAD

**Issue by slipdef** [HangfireIO/Hangfire#442]
I figured the issue is that StateChangeProcess.ChangeState method creates 2 nested connections:

![image](https://cloud.githubusercontent.com/assets/4365327/9908605/3cacc0c0-5cb6-11e5-864e-7902154db9e6.png)

It worked fine in 1.4.3

**Comment by odinserj** [comment]
Oh no, this complicates everything, need to think a lot. It is interesting, whether `TransactionScopeOption.Suppress` really suppresses a running transaction in the same connection. This will be the simplest way for running background jobs in a transaction without extending it to state changes. However, this looks like a magic and tightly related to unicorns.

Thanks for the tip!

**Comment by slipdef** [comment]
Lol. Yea, TransactionScopeOption.Suppress works fine. Previously I wrapped all Hangfire operations within TransactionScopeOption.RequiresNew scope to suppress parent transaction (all requests in our ASP.NET MVC app are wrapped within transaction scope automatically) and don't escalate to distributed transaction (we have Hangfire as a separate database). My concern with Suppress (besides that it suppresses only one parent) is that I loose transaction support for Hangfire operations.

**Comment by odinserj** [comment]
@slipdef, are you able to pass explicit connection to the `SqlServerStorage` instance?

```csharp
using (var transaction = new TransactionScope())
{
    var connection = new SqlServerConnection();
    var storage = new SqlServerStorage(connection);
    var jobClient = new BackgroundJobClient();

    jobClient.Enqueue(() => Console.WriteLine("No nested connections"));
}
```

**Comment by slipdef** [comment]
Yes, it's possible.

**Comment by odinserj** [comment]
Hangfire 1.5.0 comes with very important improvement – instant background job re-queue after process termination. This greatly reduces latency (without requiring MSMQ or RabbitMQ) and removes confusion in debug sessions, most of them end up with process termination (helpful especially for new users, since they don't know anything about invisibility timeout).

New implementation is based on regular SQL Server transactions. I've tried a lot to remove the behavior you mentioned, but each time I finished with other bad consequences:

* TransactionScope will cover user-code (job methods). This is unexpected behavior, that may lead even to deadlocks for current users.
* Increased (up to 2x) connection usage. Will lead to connection pool contention.

I'm currently thinking about third solution, but it may lead to problems too. That is why I've asked you about this consideration to have a backup plan. But is should be documented.

Btw, what problem is caused by nested connections? MSDTC?

**Comment by slipdef** [comment]
Yes, nested connection lead to MSDTC which is neither desirable in general nor possible on SQL Azure where we host Hangfire database. Passing connection explicitly looks like a fine solution for me.
